### ISSUE TITLE
S-58 Checks 1657, 1663 & 1669

### ISSUE THREAD

**Issue by iho-ohi** [iho-ohi/ENCWG-S58-ENC-Validation-Checks#15]
Proposal from nautical Dimension to improve the consistency of the tables of checks 1657, 1663 & 1669
[S58-Check 1657-1663-1669.docx](https://github.com/iho-ohi/ENCWG-S58-ENC-Validation-Checks/files/5116516/S58-Check.1657-1663-1669.docx)

**Comment by FrankHippmann** [comment]
There are some errors in the proposal document above. I have corrected these in the document attached to this comment.
[S58-Check.1657-1663-1669_FH_20200825.docx](https://github.com/iho-ohi/ENCWG-S58-ENC-Validation-Checks/files/5121017/S58-Check.1657-1663-1669_FH_20200825.docx)

**Comment by alvarosanuy** [comment]
I thinks S-58 section 1.4.3 should be amended to include new definitions: 'notPresent' and 'Any value' (see below).
All existing checks should then be reviewed for compliance. 
"Blank' spaces should not be used in order to avoid misinterpretations.

 Present – An attribute has been populated either with a value or null (255).
     notPresent - An attribute has not been populated, neither with a value nor with null (255).
 Null – An attribute has been populated with a value of null (255).
 notNull – An attribute has been populated with a value different than null (255)..
     Any Value - The attribute can be Present (populated) or notPresent (not populated).

**Comment by JeffWootton** [comment]
Tend to agree with Alvaro's suggestion.
Taking this a step further, "Unknown" is also not defined in clause 1.4.3.  Should this be "Null/notPresent" or do we need to add a separate definition for "Unknown"?

**Comment by Christian-Shom** [comment]
Agree with the proposal, although I can't see why Checks 1657 and 1663 are classified as Warning and Check 1669 as Error.
Agree with Alvaro's suggestion.
Agree with Jeff's suggestion and think "Unknown" in S-58 should be defined in line with (but could be reworded) the PS 3.5.1: "In a base data set (EN application profile), when an attribute code is present but the attribute value is missing, it means that the producer wishes to indicate that this attribute value is unknown."

**Comment by FrankHippmann** [comment]
The usage of Null within S-58 implies that it means "Unknown" for attributes. However, if we take Null to mean Unknown, then the definition for Null = 255 is incorrect. Some single byte integer subfields within ISO 8211 use 255 to signify Null, but this is not true for attribute values. Unknown (or Null) for attribute values simply means the absence of a value, not 255.

We could define Null as something like:

Null - an empty value for S-57 attributes, or 255 for ISO 8211 integer subfields

notNull could be defined as:

notNull - Present and not Null

**Comment by TDYCARHugh** [comment]
I was surprised to find that Undefined and Unknown are not really documented.
There are 3 conditions in the 8211 encoding where the  attribute record/field has 2 sub fields of code and value:

1. Undefined = Not Present = No attribute record/field is present in the data

2. Unknown = Present (there is an attribute record) with no meaningful value ( value field is NULL or 255 for ISO 8211 integers).  Unknown is really only needed when an attribute is mandatory but you don't have a value and want to indicate that it was not just overlooked.
I think that we should avoid the term Null.  It is better to use 'Unknown'.

3. Present with a meaningful value. Any value. To me 'Any Value' would not include 'Unknown' or 'Undefined'

Defined = Present (but may be Unknown)
I agree we should not use blanks in these tables, it would be better to use something like 'Not Applicable'  or 'irrelevant'.

**Comment by FrankHippmann** [comment]
If we decide to not to use Null for attribute values, we should also not use notNull. "Known" could be used instead.

Known - The attribute is present and has been populated with a value.

We need to be careful, however. If we make this change, then 106 checks that contain Null and notNull will need to be reworded. For example, 508b would change from:

For each feature object where COLPAT is notNull AND COLOUR is Null OR only has one value.

to

For each feature object where COLPAT is Known AND COLOUR is Unknown OR only has one value.

**Comment by FrankHippmann** [comment]
I correct my earlier comment, "Null - ... 255 for ISO 8211 integer subfields".

255 only represents Null for 8-bit unsigned integers. The distinction is important, e.g. RCID is a 32-bit unsigned integer, and 255 is a valid value and should not be interpreted as Null.

**Comment by LARRIVEEJCANADA** [comment]
I am agree with Hugh's comment totally, preferable to use terms used into the data providers softwares.

**Comment by FrankHippmann** [comment]
[S58-Section-1.4.3-and-Checks.1657-1663-1669.docx](https://github.com/iho-ohi/ENCWG-S58-ENC-Validation-Checks/files/5828681/S58-Section-1.4.3-and-Checks.1657-1663-1669.docx)

I have attached a revision of the proposal based on recent comments.

The proposal adds definitions for:
Known
Unknown
Optional

Note, that I have kept the terms Present and not Present, since they are already widely used throughout S-58. We should use either

Present - not Present
or
Defined - Undefined

but the mix of

Present - Undefined

is not intuitive to me.
